eZ/Bundle/EzPublishRestBundle/Tests/Functional/ContentTest.php

```diff
@@ -415,6 +415,96 @@ public function testCreateView()
         self::assertHttpResponseHasHeader($response, 'Location');
     }
 
+    /**
+     * Covers DELETE /content/objects/<contentId>/translations/<languageCode>.
+     */
+    public function testDeleteTranslation()
+    {
+        // create independent Content
+        $content = $this->createContentDraft(
+            '/api/ezp/v2/content/types/1',
+            '/api/ezp/v2/content/locations/1/2',
+            '/api/ezp/v2/content/sections/1',
+            '/api/ezp/v2/user/users/14',
+            [
+                'name' => [
+                    'eng-GB' => $this->addTestSuffix(__FUNCTION__),
+                ],
+            ]
+        );
+        $restContentHref = $content['_href'];
+        $restContentVersionHref = "{$content['Versions']['_href']}/{$content['currentVersionNo']}";
+        $this->publishContentVersionDraft($restContentVersionHref);
+        $restContentVersionHref = $this->createDraftFromVersion($content['CurrentVersion']['_href']);
+
+        // create pol-PL Translation
+        $translationToDelete = 'pol-PL';
+        $this->createVersionTranslation($restContentVersionHref, $translationToDelete, 'Polish');
+        $this->publishContentVersionDraft($restContentVersionHref);
+
+        // delete Translation
+        $response = $this->sendHttpRequest(
+            $this->createHttpRequest('DELETE', "{$restContentHref}/translations/{$translationToDelete}")
+        );
+        self::assertHttpResponseCodeEquals($response, 204);
+
+        // check that deleted Translation no longer exists
+        $response = $this->sendHttpRequest(
+            $this->createHttpRequest('GET', "$restContentHref/versions", '', 'VersionList+json')
+        );
+        self::assertHttpResponseCodeEquals($response, 200);
+        $versionList = json_decode($response->getContent(), true);
+        foreach ($versionList['VersionList']['VersionItem'] as $versionItem) {
+            self::assertNotContains($translationToDelete, $versionItem['VersionInfo']['languageCodes']);
+            foreach ($versionItem['VersionInfo']['names']['value'] as $name) {
+                self::assertNotEquals($translationToDelete, $name['_languageCode']);
+            }
+        }
+
+        return $restContentHref;
+    }
+
+    /**
+     * Covers DELETE /content/objects/<contentId>/translations/<languageCode>.
+     *
+     * @depends testDeleteTranslation
+     *
+     * @param string $restContentHref
+     */
+    public function testDeleteMainTranslation($restContentHref)
```

Comment:
This doesn't test deleting the main translation: the tested version has `pol-PL` as the main language, and `eng-GB` as a secondary one. Shouldn't this test that you get a 409 when you delete the main translation ?

As a bonus, I'd maybe reflect that in the method's name: `testDeleteMainTranslationThrows409`.

Reply:
> This doesn't test deleting the main translation: the tested version has pol-PL as the main language, and eng-GB as a secondary one. Shouldn't this test that you get a 409 when you delete the main translation ?

Hmm, I see the problem with naming. The point of this test was to check if after changing main Translation (which is a requirement), deleting previously-main Translation is still possible.

REST endpoint implements facade for working with Content item which has archived Versions having only one Translation, the one being deleted. This test was meant to test this (see also next answer about the facade).

@bdunogier So maybe if I rename it to `testDeleteTranslationDeletesSingleTranslationArchivedVersions` it would be more clear?

> As a bonus, I'd maybe reflect that in the method's name: testDeleteMainTranslationThrows409.

I could add tests testing specified HTTP errors are thrown (but for consistency more than just 409).

Reply:
> if I rename it to testDeleteTranslationDeletesSingleTranslationArchivedVersions it would be more clear?

Honestly, no, not really :)
Let me think about this.

Reply:
> The point of this test was to check if after changing main Translation (which is a requirement), deleting previously-main Translation is still possible.

In my opinion, this is why it is complicated. You are testing two use-cases at once. I'll continue as a reply to https://github.com/ezsystems/ezpublish-kernel/pull/2065#discussion_r151949770 ;)